arch: arc: archs using ATOMIC_OPERATIONS_C

IRISZZW opened the pull request:
ATOMIC_OPERATIONS_BUILTIN still has some problem in mwdt toolchain,
mwdt builtin atomic_nand() function will get wrong result.
https://github.com/zephyrproject-rtos/zephyr/blob/a06884a4e67835e57452eda348a2b77e159d4777/tests/kernel/common/src/atomic.c#L134-L138
So choosing ATOMIC_OPERATIONS_C instead.

ruuddw reviewed (changes requested):
Can you please make the change conditional to using mwdt?
(and if not done yet, create a zephyr issue and internal MetaWare ticket to get it fixed?)

IRISZZW commented:
> Can you please make the change conditional to using mwdt?
> (and if not done yet, create a zephyr issue and internal MetaWare ticket to get it fixed?)

done, internal jira number: `P10019563-43273`

ruuddw reviewed (approved):
Thanks!

ruuddw commented:
@evgeniy-paltsev is this still required for mwdt? if so, let's remove the 'stale' and get it merged.

abrodkin commented:
@evgeniy-paltsev it is still needed, so we need this one to be merged. Moreover I'd strongly suggest to mention internal bug reference `P10019563-43273` in both commit message and comments in the code so it is not lost.

ruuddw commented:
@IRISZZW Can you please fix the CI checks so it can be merged?
Or, do we want to simply ignore this issue till the builtin gets fixed (so SMP doesn't pay the performance penalty of the generic C implementation? Add a custom implementation that uses correct assembly implementation instead of the built-in?

abrodkin commented:
Well, actually we might want to disable usage of MWDT for SMP platforms until we get fixed MetaWare's built-in atomics, because atomics implemented in C (as they are now) are not really ATOMIC on SMP systems :)

Moreover C-implemented atomics are fundamentally broken for SMP systems because they lead to an infinite loop in `k_spin_lock()` that funny way:
1. `atomic_cas()` (see https://elixir.bootlin.com/zephyr/latest/source/kernel/atomic_c.c#L81) calls `k_spin_lock()`
2. `k_spin_lock()` in case of `CONFIG_SMP` calls `atomic_cas()`
3. Back to (1) :)

So what needs to be done is for `CONFIG_ATOMIC_OPERATIONS_C` dependency on `!CONFIG_SMP` to be added.
And that will implicitly disable MWDT for HSDK (until it is configured with SMP).

evgeniy-paltsev commented:
I propose to have something like that
```
select ATOMIC_OPERATIONS_C if $(ZEPHYR_TOOLCHAIN_VARIANT) == "arcmwdt" && !SMP
select ATOMIC_OPERATIONS_BUILTIN if $(ZEPHYR_TOOLCHAIN_VARIANT) != "arcmwdt" || SMP
```

until we get atomics in MWDT fixed (or until we create our custom implementation)
That will allow us to get better results at least on UP platforms with MWDT.

ruuddw commented:
If the generic C atomics don't work for SMP at all, I'd opt for Alexey's suggestion.

evgeniy-paltsev commented:
> If the generic C atomics don't work for SMP at all, I'd opt for Alexey's suggestion.

That is also fine for me :)

abrodkin reviewed (changes requested):
`ATOMIC_OPERATIONS_C ` are even in theory are not compatible with `SMP` so it needs to be fixed in a generic way as it has nothing to do with any given architecture.

I.e. `select ATOMIC_OPERATIONS_C if ZEPHYR_TOOLCHAIN_VARIANT = "arcmwdt"` makes sense here, but SMP part needs to be addressed elsewhere.

IRISZZW commented:
> `ATOMIC_OPERATIONS_C ` are even in theory are not compatible with `SMP` so it needs to be fixed in a generic way as it has nothing to do with any given architecture.
> 
> I.e. `select ATOMIC_OPERATIONS_C if ZEPHYR_TOOLCHAIN_VARIANT = "arcmwdt"` makes sense here, but SMP part needs to be addressed elsewhere.

I will create a PR for the dependence between ATOMIC_OPERATIONS and SMP, and keep this PR focus on ZEPHYR_TOOLCHAIN_VARIANT and ATOMIC_OPERATIONS, it's OK?

evgeniy-paltsev commented on arch/arc/Kconfig:
```diff
@@ -9,6 +9,12 @@ menu "ARC Options"
 config ARCH
 	default "arc"
 
+config ZEPHYR_TOOLCHAIN_VARIANT
+	string "ZEPHYR_TOOLCHAIN_VARIANT"
+	default "$(ZEPHYR_TOOLCHAIN_VARIANT)"
+	help
+	  zephyr_toolchain_variant get from env
+
```
Any explanation about why do we need this?

ZEPHYR_TOOLCHAIN_VARIANT is used directly in other places:
```
config TOOLCHAIN_SUPPORTS_THREAD_LOCAL_STORAGE
	bool
	default y if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr"
	help
	  Hidden option to signal that toolchain supports generating code
	  with thread local storage.
```

IRISZZW replied:
Thanks for your point, I should use `"$(ZEPHYR_TOOLCHAIN_VARIANT)"`, not` $(ZEPHYR_TOOLCHAIN_VARIANT)`

IRISZZW committed:
arch: arc: archs using ATOMIC_OPERATIONS_C

ATOMIC_OPERATIONS_BUILTIN still has some problem in mwdt toolchain,
so choosing ATOMIC_OPERATIONS_C instead.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
